packages/flutter_tools/lib/src/android/android_device.dart

```diff
@@ -90,8 +101,17 @@ class AndroidDevice extends Device {
   @override
   Future<bool> get isLocalEmulator async {
     if (_isLocalEmulator == null) {
-      final String characteristics = await _getProperty('ro.build.characteristics');
-      _isLocalEmulator = characteristics != null && characteristics.contains('emulator');
+      final String hardware = await _getProperty('ro.hardware');
+      printTrace('ro.hardware = $hardware');
+      if (_knownHardware.containsKey(hardware)) {
+        // Look for known hardware models.
+        _isLocalEmulator = _knownHardware[hardware] == _HardwareType.emulator;
+      } else {
+        // Fall back to a best-effort heuristic-based approach.
+        final String characteristics = await _getProperty('ro.build.characteristics');
+        printTrace('ro.build.characteristics = $characteristics');
+        _isLocalEmulator = characteristics != null && characteristics.contains('emulator');
```

Comment:
As we discussed, my understanding is this may be the wrong question for the tool to ask?

What changes our engine config is:
1.  Does the system have a GPU (if not, use --debug, software backend).
2.  Do we have a --release binary for this architecture (x86, x64, etc.) . If not force --debug.